### PR TITLE
Python: always use --no-build-isolation

### PR DESCRIPTION
**Pull request by singingwolfboy**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

Follow-up to #15121. Now that https://github.com/Homebrew/homebrew-core/pull/125779 is merged, validating the need for this `--no-build-isolation` flag, we can consider passing this flag all the time. I think it's probably a good idea, but I'm certainly open to discussion!

Once this pull request merges, I'll make a PR to `homebrew-core` to remove the usage of the `build_isolation: false` param from the `pytorch` and `torchvision` formulae.

### CODE REVIEW

**Review by MikeMcQuaid — APPROVED** [review]
Makes sense to me, thanks @singingwolfboy! Will hold off merging for now in case any @Homebrew/core folks have opinions.

**Review by carlocab — APPROVED** [review]
> Once this pull request merges, I'll make a PR to `homebrew-core` to remove the usage of the `build_isolation: false` param from the `pytorch` and `torchvision` formulae.

Usage probably needs to be removed first before merging this. But, otherwise, ✅

**Comment by singingwolfboy** [comment]
> Usage probably needs to be removed first before merging this. But, otherwise, ✅

It's a chicken-and-egg problem, unfortunately. The formulae won't build without the --no-build-isolation flag, so I can't remove the usage until this flag is passed automatically.

**Comment by dtrodrigues** [comment]
Using `--no-build-isolation` would be preferable for the reasons listed in https://github.com/Homebrew/brew/pull/10811 , but we'd also need to provide all build-time dependencies since `pip` now won't download them. Formulae that use `hatchling`, `poetry-core`, `flit` etc to build will need to be updated at a minimum, so I'm not sure if this is the correct default. For instance, with this PR, ` brew install -s pywhat` now will return
```
==> /opt/homebrew/Cellar/pywhat/5.1.0_1/libexec/bin/pip install -v --no-deps --no-binary :all: --use-feature=no-binary-enable-wheel-cache --ignore-installed --no-build-isolation /private/tmp/pywhat--color
Last 15 lines from /Users/dustin/Library/Logs/Homebrew/pywhat/03.pip:
  File "/opt/homebrew/Cellar/python@3.11/3.11.3/Frameworks/Python.framework/Versions/3.11/lib/python3.11/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap>", line 1206, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1178, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1128, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "<frozen importlib._bootstrap>", line 1206, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1178, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1142, in _find_and_load_unlocked
ModuleNotFoundError: No module named 'hatchling'
```

**Comment by singingwolfboy** [comment]
Thanks for pointing that out, @dtrodrigues! Now that I've read through #10811, I see that there's a lot of work to be done before making this the default.

Maybe a good course of action would be to first migrate all Python formulae to use this `build_isolation: false` flag, and then remove the option? That seems like a lot of work, but it might be the best course of action. What do you think?

**Comment by dtrodrigues** [comment]
@alebcay did a lot of good work in #10811, but as you observed, there's still a significant amount of work required before arbitrary Python formulae build with `--no-build-isolation`. As I mentioned in https://github.com/Homebrew/brew/pull/10811#issuecomment-794614941, in an ideal world I'd also like to be able to pass `--no-index` to prevent any downloads from PyPI that we're not explicitly downloading in a `resource` block from being used. This would be major change from how Python formulae currently work in terms of how the formulae need to be written/updated, as `PyPI.update_python_resources` focuses only on runtime dependencies right now. There's also the potential to have build dependencies that conflict with each other or with other runtime dependencies.

For now, I think it makes sense to only pass `--no-build-isolation` in the cases like pytorch/torchvision that require it.

**Comment by MikeMcQuaid** [comment]
Passing on this for now given above.

**Pull request closed without merging**
